### PR TITLE
Enable LG debuggability

### LINKED ISSUE #2385: [Adaptive] Adaptive dialogs + LG debuggability
Work here builds on top of the inspection middleware to enable improved debugging experience with adaptive dialogs.

1. Add support for ability in the runtime to break at a specific step in an adaptive dialog
2. Add ability to continue to next step and break once a break step has been encountered. 
3. Add ability to continue till next break point (if any) once a break step has been encountered.
4. Add debug break property for all steps and events.

### PR DESCRIPTION
**Pull request by Danieladu**
fixes #2385 
## Description
Enable LG debuggability with event handler.
1. Add LG event and Expression event args in LG sdk
2. Register LG/Expression object and source range
3. Emit event during LG evaluation and Expression evaluation
4. Handler event in the adaptive language generator

## How to detect the debugger
1. Set up an Adaptive dialog bot, and startup the project
2. Install adaptive-tool extension for vscode. (Currently, adaptive-tool extension could be built from [source code](https://github.com/microsoft/BotBuilder-Samples/tree/master/experimental/adaptive-tool))
3. Goto the LG file that the adaptive-bot used, and set up the [launch file](https://github.com/microsoft/BotBuilder-Samples/tree/master/experimental/adaptive-tool#setting-up), press F5 to listen on the bot service runtime
4. Insert breaking points for Lg templates or Expressions
5. Go through the bot flow, if the bot has reached a certain template or a certain expression, the code will stop at the breakpoint.
Check the variables and status to get more details.

## Details for this PR
- Add four event args class: LGEventArgs, MessageArgs, BeginTemplateEvaluationArgs, and BeginExpressionEvaluationArgs.
They all inherit from LGEvent which is registered in dialogEvents.
- Add ExpressionRef class as the breakpoint matching object, since the debugger uses "reference equal" as the breaking point equivalence.
- Register the source map of templates and expressions during the paring stage.
- Add the event handler at the Adaptive generator layer. When there is LG/Expression hitting event, check the debugger to see if there is a breaking point at the current line. 
- Emit LG event when the code is hitting a certain template during the evaluation stage, and also emit the expression event when the code is evaluating a certain expression.
- Adaptive dialog would subscribe to these events and emit [dap](https://microsoft.github.io/debug-adapter-protocol/) messages to vscode.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs [hunk 1]
```diff
@@ -186,16 +193,16 @@ private static IParseTree AntlrParseTemplates(string text, string id)
             return parser.file();
         }
 
-        private static IList<Templates> GetReferences(Templates file, Dictionary<string, Templates> cachedTemplates = null)
+        private static IList<Templates> GetReferences(Templates file, Dictionary<string, Templates> cachedTemplates = null, EventHandler registeSourceMap = null)
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs:196]: I'm not sure why we need eventCallback in parsing process? 

I assume people can take the result of parseTree, and do what ever they want, so event callback is more useful in evaluation pharse than parse phase, isn't it?

Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs:196]: I know, you mean we provide a new method `registeTemplateAndExpression(EventHandler onEvent)`, in this method, we would go through the parsetree and register all templates and expressions.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 2]
```diff
@@ -67,6 +70,7 @@ public TemplateEngineLanguageGenerator(string filePath, Dictionary<string, IList
             var (_, locale) = LGResourceLoader.ParseLGFileName(Id);
             var importResolver = LanguageGeneratorManager.ResourceExplorerResolver(locale, resourceMapping);
             this.lg = LanguageGeneration.Templates.ParseFile(filePath, importResolver);
+            this.lg.AddDebuggingEventRegister(SourceMapRegister);
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:73]: i don't think we should require a callback just for registering sourcemap, a callback function should only be required when do evalutation. 

If certain sub-structure is not not available after parseFile, we should extend that, to make it available. 

The principle is after parsing, i should have enough information to register source map.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 3]
```diff
@@ -67,6 +69,7 @@ public TemplateEngineLanguageGenerator(string filePath, Dictionary<string, IList
             var (_, locale) = LGResourceLoader.ParseLGFileName(Id);
             var importResolver = LanguageGeneratorManager.ResourceExplorerResolver(locale, resourceMapping);
             this.lg = LanguageGeneration.Templates.ParseFile(filePath, importResolver);
+            AddDebuggingEventRegister(lg);
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:72]: rename this to registerLGTemplates make more sense to me, or registerSourceMap

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs [hunk 4]
```diff
@@ -10,5 +10,8 @@ public class DialogEvents
         public const string CancelDialog = "cancelDialog";
         public const string ActivityReceived = "activityReceived";
         public const string Error = "error";
+        public const string BeginTemplateEvaluation = "beginTemplateEvaluation";
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs:13]: is it possible to move those two events into maybe LGEvents?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs [hunk 5]
```diff
@@ -443,12 +456,52 @@ private object EvalExpression(string exp, ParserRuleContext context = null, stri
...
+                    if (string.IsNullOrEmpty(result.error))
+                    {
+                        text = $"Evaluate expression '{exp}' get result: {result.value}";
+                    }
+                    else
+                    {
+                        text = $"Evaluate expression '{exp}' get error: {result.error}";
+                    }
+
+                    lgOptions.OnEvent(MessageKey, new MessageArgs { Source = source, Text = text, Context = currentTemplate.SourceRange.ParseTree });
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs:500]: what is this MessageKey used for?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionContext.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
+﻿namespace Microsoft.Bot.Builder.LanguageGeneration
+{
+    /// <summary>
+    /// Expression container with source range.
+    /// </summary>
+    public class ExpressionContext
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionContext.cs:6]: ExpressionRef or ExpressionContext? ExpressionContext gives more feeling like this is a context used to evaluate expression?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Template.cs [hunk 7]
```diff
@@ -31,8 +31,17 @@ public class Template
...
+            Expressions = new List<ExpressionContext>();
         }
 
+        /// <summary>
+        /// Gets or sets expression context list.
+        /// </summary>
+        /// <value>
+        /// Expression context list.
+        /// </value>
+        public IList<ExpressionContext> Expressions { get; set; }
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Template.cs:43]: I prefer consistent naming pattern, if you call this ExpressionContext expression, i think this indicate ExpressionContext is not a good name?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs (hunk 5 above)
Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs:500]: it is a placeholder, just mark that this event is an message event.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs [hunk 8]
```diff
@@ -113,19 +114,32 @@ public object EvaluateTemplate(string inputTemplateName, object scope)
 
             // Using a stack to track the evaluation trace
             evaluationTargetStack.Push(templateTarget);
-            var result = Visit(TemplateMap[templateName].TemplateBodyParseTree);
+            currentTemplate = CurrentTemplate();
+            var source = currentTemplate.SourceRange.Source;
+
+            if (Path.IsPathRooted(source) && lgOptions.OnEvent != null)
+            {
+                lgOptions.OnEvent(currentTemplate, new BeginTemplateEvaluationArgs { Source = source, TemplateName = templateName, Context = currentTemplate.SourceRange.ParseTree });
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs:122]: this is an async function, and you are not awaiting it? how does that work?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 9]
```diff
@@ -101,5 +106,57 @@ public override async Task<string> Generate(DialogContext dialogContext, string
...
+                    task = dialogContext.GetDebugger().StepAsync(dialogContext, sender, expr.Type, new System.Threading.CancellationToken());
+                }
+                else if (e is MessageArgs message && dialogContext.GetDebugger() is IDebugger dda)
+                {
+                    task = dda.OutputAsync(message.Text, sender, message.Text, new System.Threading.CancellationToken());
+                }
+
+                if (task != null)
+                {
+                    task.Wait();
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:130]: can we use something like AsyncHelper to make sure it won't block

File: libraries/Microsoft.Bot.Builder/StringUtils.cs [hunk 10]
```diff
@@ -32,6 +33,7 @@ public static string Ellipsis(string text, int length)
         /// <returns>string which is unique SHA256 hash.</returns>
         public static string Hash(string text)
         {
+            text = text ?? string.Empty;
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder/StringUtils.cs:36]: >string.Empty [](start = 27, length = 12)

this should be 
     throw new ArgumentNull(nameof(text))

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs (hunk 4 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs]: Yes this should NOT be in this assembly

---
In reply to: [416500764](https://github.com/microsoft/botbuilder-dotnet/pull/3791#discussion_r416500764) [](ancestors = 416500764)

File: libraries/Microsoft.Bot.Builder.Dialogs.Debugging/DebuggerSourceMap.cs [hunk 11]
```diff
@@ -64,7 +64,7 @@ void ISourceMap.Add(object item, SourceRange range)
         {
             if (range.Path != null && !Path.IsPathRooted(range.Path))
             {
-                throw new ArgumentOutOfRangeException(range.Path);
+                //throw new ArgumentOutOfRangeException(range.Path);
             }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Debugging/DebuggerSourceMap.cs:68]: What? This can't be right.

Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs.Debugging/DebuggerSourceMap.cs]: I will remove it

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs [hunk 12]
```diff
@@ -103,6 +105,14 @@ public EvaluationOptions(IList<string> optionsList)
         /// </value>
         public Func<string, object> NullSubstitution { get; set; } = null;
 
+        /// <summary>
+        /// Gets or sets an event handler that handles the emited events in the evaluation process.
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs:109]: >emited [](start = 59, length = 6)

Should be "emitted" here and in the comment below.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 13]
```diff
@@ -157,8 +157,8 @@ public class Templates : List<Template>
         /// Parser to turn lg content into a <see cref="LanguageGeneration.Templates"/>.
         /// </summary>
         /// <param name="filePath">Absolute path of a LG file.</param>
-        /// <param name="importResolver">resolver to resolve LG import id to template text.</param>
-        /// <param name="expressionParser">expressionEngine Expression engine for evaluating expressions.</param>
+        /// <param name="importResolver">Resolver to resolve LG import id to template text.</param>
+        /// <param name="expressionParser">ExpressionEngine Expression engine for evaluating expressions.</param>
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:161]: >ExpressionEngine [](start = 43, length = 16)

Drop ExpressionEngine--it is not the type anyway.

File: tests/Microsoft.Bot.Builder.TestBot.Json/Samples/ToDoBot/.vscode/launch.json [hunk 14]
```diff
@@ -5,7 +5,7 @@
     "version": "0.2.0",
     "configurations": [
         {
-            "type": "json",
+            "type": "bot",
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.TestBot.Json/Samples/ToDoBot/.vscode/launch.json:8]: >bot [](start = 21, length = 3)

Is that an actual type?

Comment (Danieladu) [reply tests/Microsoft.Bot.Builder.TestBot.Json/Samples/ToDoBot/.vscode/launch.json:8]: yes, it is. The type name is the same with the type name of adaptive-tool.
Originally, ".dialog" is the only debug target, so the type name is "json"(.dialog's default schema format) 
Now, we add LG into the debugging tool, So, type "json" should be dropped.

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs (hunk 4 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs]: This is a REQUIRED change, you cannot put this here.

---
In reply to: [417385145](https://github.com/microsoft/botbuilder-dotnet/pull/3791#discussion_r417385145) [](ancestors = 417385145,416500764)

**Comment by sgellock** [comment]
moving this out of R9

**Comment by sgellock** [comment]
This should not be merged until we open master for R10 commits

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/BeginExpressionEvaluationArgs.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Microsoft.Bot.Builder.LanguageGeneration
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/BeginExpressionEvaluationArgs.cs:1]: Missing //Copyright headers

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/BeginTemplateEvaluationArgs.cs [hunk 16]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Microsoft.Bot.Builder.LanguageGeneration
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/BeginTemplateEvaluationArgs.cs:1]: //Coppyright

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionRef.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
+﻿namespace Microsoft.Bot.Builder.LanguageGeneration
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionRef.cs:1]: Copyright

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionRef.cs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        /// </value>
+        public SourceRange SourceRange { get; set; }
+
+        public override string ToString()
+        {
+            return Expression;
+        }
+
+        /// <summary>
+        /// Get the unique id of erxpression context.
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionRef.cs:36]: typo on expression

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionRef.cs [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+        /// </summary>
+        /// <returns>id string.</returns>
+        public string GetId()
+        {
+            return SourceRange?.Source + ":" + SourceRange?.Range + ":" + Expression;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/ExpressionRef.cs:41]: if sourcerange is null, this won't return a unique Id.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/LGEventArgs.cs [hunk 20]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/LGEventArgs.cs:1]: // Coppyright

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/MessageArgs.cs [hunk 21]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace Microsoft.Bot.Builder.LanguageGeneration
+{
+    public class MessageArgs : LGEventArgs
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Events/MessageArgs.cs:5]: Each of these 1-off classes needs comments, and a test class.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 22]
```diff
@@ -103,5 +115,58 @@ public override Task<object> GenerateAsync(DialogContext dialogContext, string t
...
+                    await dialogContext.GetDebugger().StepAsync(dialogContext, sender, eventType, new CancellationToken()).ConfigureAwait(false);
+                }
+                else if (e is MessageArgs message && dialogContext.GetDebugger() is IDebugger dda)
+                {
+                    await dda.OutputAsync(message.Text, sender, message.Text, new CancellationToken()).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private void RunSync(Func<Task> func)
```
[the remark this answers is not in the document]
Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs]: It is recommended by tom. I have made it static.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 23]
```diff
@@ -103,5 +115,58 @@ public override Task<object> GenerateAsync(DialogContext dialogContext, string t
...
+        {
+            var culture = CultureInfo.CurrentCulture;
+            TaskFactory.StartNew(() =>
+            {
+                Thread.CurrentThread.CurrentCulture = culture;
```
[the remark this answers is not in the document]
Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs]: Just remove the culture info. Have any solutions? I have no idea about this

**Comment by Danieladu** [comment]
@cleemullins  @tomlm  may you take a look and to see if there is any problem.

**Comment by Danieladu** [comment]
blocked by the ID discussion https://github.com/microsoft/botbuilder-dotnet/pull/4155#discussion_r446528210

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs [hunk 24]
```diff
@@ -36,5 +36,10 @@ public class DialogEvents
         /// Event fired when there was an exception thrown in the system.
         /// </summary>
         public const string Error = "error";
+
+        /// <summary>
+        /// Event fired by LG library.
+        /// </summary>
+        public const string LGEvents = "lgEvents";
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs:43]: this is a layer violation.  This constant should simply be in LAnguageGenerationAssembly in a class called LGDialogEvents

Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs:43]: But if the event is not registered in the Dialog package, the breaking points would not work.
ref: the original source code:
https://github.com/microsoft/botbuilder-dotnet/blob/cdaa26ca15db8e657a3b4c7bb77a76f5e18a0c91/libraries/Microsoft.Bot.Builder.Dialogs.Debugging/DialogDebugAdapter.cs#L158

The `_events` only contains the events in DialogEvents as default.
So, should I extend the DialogEvents class and add a "lgEvent" event, and pass it into the DialogDebugAdapter in "UseDebugger" middleware?

Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs:43]: Yes, there are 2 ways to resolve this.
1. You can pass in a custom array of _events when UseDebugger() is called.  
2.  we could give this a name which is generic.  What about Custom = "custom" ? Evaluation="Evaluation"?  
I think I prefer #2.  Is there any reason this needs to be called LGEvent?

---
In reply to: [479843093](https://github.com/microsoft/botbuilder-dotnet/pull/3791#discussion_r479843093) [](ancestors = 479843093)

Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs:43]: Prefer 2 too. I would change this part soon. Thanks.

Comment (Danieladu) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogEvents.cs]: # fixed

**Comment by Danieladu** [comment]
LG debugger is ready to review. All kinds of discussions and suggestions are welcome 😊
@boydc2014  @tomlm @chrimc62 @cleemullins

**Comment by mrivera-ms** [comment]
@tomlm , @cleemullins , @boydc2014 , please review if all your concerns have been addressed. Thank you.

**Review by boydc2014 — APPROVED** [review]

**Pull request merged**